File: src/main/java/macro/library/database/IdTable.java
package macro.library.database;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Created by Macro303 on 2019-Oct-22
 */
public abstract class IdTable<T, S> extends Table<T> {
	@NotNull
	protected final String idName;

	protected IdTable(@NotNull String tableName, @NotNull String idName) {
		super(tableName);
		this.idName = idName;
	}

	@Nullable
	public T selectUnique(S unique) {
		var query = String.format("SELECT * FROM %s WHERE %s = ?", tableName, idName);
		return search(query, unique).stream().findFirst().orElse(null);
	}

	public boolean delete(S unique){
		var query = String.format("DELETE FROM %s WHERE %s = ?;", tableName, idName);
		return delete(query, unique);
	}

	public abstract boolean insert(T item);
	public abstract boolean update(T item);
}

File: src/main/kotlin/macro/library/database/IdTable.kt
package macro.library.database

import org.apache.logging.log4j.LogManager

/**
 * Created by Macro303 on 2019-Oct-30
 */
abstract class IdTable<T, S> protected constructor(tableName: String, protected val idName: String) :
	Table<T>(tableName = tableName) {

	fun selectUnique(unique: S): T? {
		val query = "SELECT * FROM $tableName WHERE $idName = ?"
		return search(unique, query = query).firstOrNull()
	}

	fun delete(unique: S): Boolean {
		val query = "DELETE FROM $tableName WHERE $idName = ?;"
		return delete(unique, query = query)
	}

	abstract fun insert(item: T): Boolean

	abstract fun update(item: T): Boolean

	companion object {
		private val LOGGER = LogManager.getLogger(IdTable::class.java)
	}
}